File: MPC/process_mask.py
import cv2
import numpy as np

def calcular_deslocamento_lateral(mascara):
    # Máscara: imagem binária 256x128, branco (255) para linhas, preto (0) para fundo
    # Pegar a linha inferior (y=128) ou próxima (y=120 para evitar bordas)
    linha = mascara[300, :]  # Linha y=120

    # Encontrar pixels brancos (linhas)
    pixels_brancos = np.where(linha == 255)[0]  # Índices dos pixels brancos

    if len(pixels_brancos) == 0:
        return 0  # Retornar 0 se não encontrar linhas

    # Linha esquerda: menor x; linha direita: maior x
    x_esquerda = pixels_brancos.min()
    x_direita = pixels_brancos.max()

    # Calcular centro
    x_centro = (x_esquerda + x_direita) / 2

    # Centro da imagem (alinhado com o veículo)
    x_centro_imagem = 640 / 2  # 128 pixels

    # Deslocamento lateral (em pixels)
    deslocamento = x_centro - x_centro_imagem

    print(f"Linha esquerda: x={x_esquerda}, Linha direita: x={x_direita}")
    print(f"Centro da trajetória: x={x_centro}")
    print(f"Deslocamento lateral: {deslocamento} pixels")

    return deslocamento

# Exemplo de uso
mascara = cv2.imread('mask/mask_test01.png', cv2.IMREAD_GRAYSCALE)  # Carregar máscara
deslocamento = calcular_deslocamento_lateral(mascara)
print(f"Deslocamento lateral: {deslocamento} pixels")

File: MPC/process_mask.cpp
#include <opencv2/opencv.hpp>
#include <iostream>
#include <vector>

int h = 480;
int x = 640;

float calLatShift(const cv::Mat& mask){
    // mask: imagem binária 256x128
    // line y=120 (para evitar bordas)
    cv::Mat line = mask.row(479);

    // find white pixels(lines)
    std::vector<int> whitePixels;
    for (int x = 0; x < line.cols; ++x){
        if (line.at<uchar>(0, x) == 255)
            whitePixels.push_back(x);
    }


    // check if find lines
    if (whitePixels.empty()){
        std::cout << "None line found in mask" << std::endl;
        return 0.0f;
    }

    // left line: minor x; right line: larger x
    int xLeft = *std::min_element(whitePixels.begin(), whitePixels.end());
    int xRight = *std::max_element(whitePixels.begin(), whitePixels.end());

    // center cal
    float xCenter = (xLeft + xRight) / 2.0f;

    // Center of img (aligned with the vehicle)
    float xCenterImg = x / 2.0f;

    // Shift lateral (pixels)
    float shift = xCenter - xCenterImg;

    // Imprimir informações para depuração
    std::cout << "Linha esquerda: x=" << xLeft << ", Linha direita: x=" << xRight << std::endl;
    std::cout << "Centro da trajetória: x=" << xCenter << std::endl;
    std::cout << "Deslocamento lateral: " << shift << " pixels" << std::endl;

    return shift;
}

int main () {
    // Carregar máscara (exemplo: substitua pelo caminho da sua máscara)
    cv::Mat mask = cv::imread("mask/mask_test01.png", cv::IMREAD_GRAYSCALE);
    
    if (mask.empty()) {
        std::cerr << "Erro ao carregar a máscara!" << std::endl;
        return -1;
    }

    // Verificar dimensões
    if (mask.cols != x || mask.rows != h) {
        std::cerr << "Máscara com dimensões incorretas! Esperado: 256x128" << std::endl;
        return -1;
    }

    // Calcular deslocamento
    float shift = calLatShift(mask);

    return 0;
}